content/posts/2022-02-02-decreased-backend-tests-step/index.md

```diff
@@ -0,0 +1,174 @@
+
+In this particular initiative, the team was formed by 3 full-stack engineers, 2 backends end engineers, and 1 one site reliability engineer, all of them from different squads.
+
+## Goals
+As with all Loadsmart initiatives, it should have clear goals and be based on data to be properly tracked by everyone.
+
+For this specific effort, we came across the following goals:
+
+- Reduce average execution time of backend tests by 50%
+    - How to measure: CircleCi backend tests step execution time
+
+- Reduce average time from code-ready to code-deployed by 50%
+    - How to measure: CircleCi backend tests step + QA deploy + Staging deploy + Production deploy
+
+
+## Methodology
+
+We followed a scientific method for problem-solving based on the following steps:
+
+- Identify a problem:
+We may have several potential issues in our code impacting the backend test step execution time.
+However, we should pick one to focus on. In this step, we took several problems and guessed which one of them would be better to start with.
+
+- Research:
+After having a clear definition of the problem, we started doing some research to understand how it was being handled by the software community.
+
+- Hypothesis:
+Based on the results of previous steps, we could gather some hypotheses in the format of "what if...".
+
+- Experimentation:
+For each hypothesis, we have evaluated the effort to test it. After qualifying it as a valid one, we prototyped it.
+We did it for a minimal set of changes enabling an initial evaluation. Our goal here was to change minimal and get the max result possible.
+
+- Results and conclusions:
+Once we had a prototype working as expected, we started gathering data for it. If it showed good results, we could invest more effort in it.
+
+## Actions and Results
+
+### Baseline
+
+As a baseline, we considered the total execution time for the backend test step: 22 minutes.
+
+1. Map and split slowest Django apps
+
+We had a bunch of tests spread through several Django apps.
+
+We came up with the following hypothesis: What if we could split the faster/slower apps into different jobs?
+
+We started by measuring the execution time for each Django app. Then, we created manually 2 test jobs: the first would have the slowest tests and the second would have the fastest. We tried to split it equally based on total execution time, ie, the second job with faster tests would have more tests than the first.
+
+We got the following results showing an improvement of about 7 minutes
+
+- Job 1 - 14 min 35 s
+- Job 2 - 14 min 15 s
+
+It showed a valid path which was splitting tests to run in parallel. However, the solution to split them into multiple jobs was not feasible, since we're keeping adding more and more tests daily.
```

Comment:
```suggestion
It showed a valid path which was splitting tests to run in parallel. However, the solution to split them into multiple jobs was not feasible, since we'd keep adding more and more tests daily.
```